### ISSUE TITLE
Move defining query implementation to InMemory

### ISSUE THREAD

**Issue by AndriySvyryd** [dotnet/efcore#21624]
After the methods obsoleted in #18903 are removed the implementation and tests can be made in-memory-only.

**Comment by smitpatel** [comment]
This cannot be done in 5.0

**Comment by smitpatel** [comment]
@dotnet/efteam - Defining queries are registered in model in OnModelCreating. Should it disallow usage of DbContext based members other than DbSet? Currently it may be working by accident in the way how they work. But we are functionally breaking it to make defining query work like a view for InMemory. Database views doesn't access DbContext members.

**Comment by AndriySvyryd** [comment]
Yes, now would be a good moment to disallow it so we can get feedback from early adopters.

**Comment by smitpatel** [comment]
Few notes:
- InMemory has TPH like mapping (so hierarchy must have discriminator property) but stored like TPC so each entity type has own InMemoryTable. Similar fashion, we will allow each entity type to have own defining query. For types which doesn't have it in hierarchy we will fall back to the table. We do union across all tables in InMemoryStore. So similar fashion union should be possible across mixed mapping.
- Query pipeline needs to generate stream of ValueBuffer to be consumed just like table. We need to utilize query pipeline to do so. Note to ignore query filter when generating this so entity type level query filter should filter out results as needed in actual query. There could be dependency issues between which defining query needs to be processed first.
- Above processing can go into model finalization step and store a delegate to generate stream in actual query.
- Log a warning for any type in hierarchy not mapped to defining query if any of them is. (so mixed mapping is noticed by user).
- Block invalid operations inside defining query like Include/IgnoreQueryFilter/AsTracking* methods.
- The output of defining query can be a concrete type which is exactly same type as the entityType CLR type or an anonymous type which member names matching property names (anonymous type will allow providing values for shadow properties).
